embedded-hal-async/src/serial.rs

```diff
@@ -0,0 +1,68 @@
+//! Serial interface
+
+use core::future::Future;
+pub use embedded_hal::serial::{Error, ErrorKind, ErrorType};
+
+/// Read half of a serial interface
+///
+/// Some serial interfaces support different data sizes (8 bits, 9 bits, etc.);
+/// This can be encoded in this trait via the `Word` type parameter.
+pub trait Read<Word: 'static + Copy = u8>: ErrorType {
+    /// Future returned by the `read` method.
+    type ReadFuture<'a>: Future<Output = Result<(), Self::Error>> + 'a
+    where
+        Self: 'a;
+
+    /// Reads words from the serial interface into the supplied slice.
+    fn read<'a>(&'a mut self, read: &'a mut [Word]) -> Self::ReadFuture<'a>;
```

Comment:
per discussion on the previous PR, i think we need to document whether this requires the buffer to be filled or can return partial results on internal failure.

i'd lean towards the linux-esque "may return early" so implementations can break on End of Frame if they have the capability, and return partial data where it is reasonable to do so, perhaps with a `read_all` or `read_exact` function if you expect the whole buffer to be filled (which i think can be easily implemented on the modified `read` function).

Reply:
@Dirbaio (@eldruin) thoughts on adding a `usize` to `ReadFuture` and allowing (but not requiring) early returns? that way this will play well with linux or the nrf, and to ensure a full buffer at the application level you just stack read calls.

Reply:
I don't think allowing different impls to behave differently is a good idea. The trait should specify which behavior it is, and all impls should do that.

If some impls do early returns and some don't, you can have scenarios like a driver that needs it works fine on Linux, and then breaks on nrf.